Gems/Profiler/Code/Source/ImGuiTreemap.cpp

```diff
@@ -0,0 +1,645 @@
+        AZStd::vector<TreemapNode*> children;
+        children.reserve(node.m_children.size() + 1);
+        for (TreemapNode& child : node.m_children)
+        {
+            if (child.m_tag == 0 || (m_currentMask & child.m_tag) > 0)
+            {
+                children.push_back(&child);
+            }
+        }
+
+        // This dummy node at the end is needed to finalize the last row (which will be the last child node
+        // occupying the row by itself).
+        TreemapNode endSentinel;
+        endSentinel.m_weight = 0.f;
+        children.push_back(&endSentinel);
+
+        AZStd::sort(
+            children.begin(), children.end(),
+            [](const TreemapNode* lhs, const TreemapNode* rhs)
+            {
+                return lhs->m_weight > rhs->m_weight;
+            });
+
+        // Shrink the frame corresponding to a node to ensure there's a 2 pixel gutter.
+        int rowExtent[2] = { node.m_extent[0] - 4, node.m_extent[1] - 4 };
+        int rowOffset[2] = { node.m_offset[0] + 2, node.m_offset[1] + 2 };
+        bool horizontal = rowExtent[1] > rowExtent[0];
+        // The "extent" here tracks the extent along the rows orientation. The row we lay out entries within could be oriented vertically
+        // depending on the aspect ratio.
+        int extent = horizontal ? rowExtent[0] : rowExtent[1];
+
+        AZStd::vector<TreemapNode*> row;
+        double rowArea = 0.f;
+        double rowMinArea = 0.f;
+        double rowMaxArea = 0.f;
+        double grade = 0.f;
+
+        // The aspect ratios are computed with respect to element areas, so compute those areas here
+        double scale = rowExtent[0] * rowExtent[1] / node.m_weight;
+        for (TreemapNode* child : children)
+        {
+            child->m_area = child->m_weight * scale;
+        }
+
+        for (auto it = children.begin(); it != children.end(); /* iterator conditionally incremented in loop body */)
+        {
+            TreemapNode& child = **it;
+
+            // If the row is empty, unconditionally start a new row.
+            if (row.empty())
+            {
+                row.push_back(&child);
+                rowArea = child.m_area;
+                rowMinArea = child.m_area;
+                rowMaxArea = child.m_area;
+                grade = GradeRow(rowArea, rowMinArea, rowMaxArea, extent);
+                ++it;
+                continue;
+            }
+
+            // Check if this node belongs in the current row, or if we should start a new one
+            double gradeIfAdded =
+                GradeRow(rowArea + child.m_area, AZStd::min(child.m_area, rowMinArea), AZStd::max(child.m_area, rowMaxArea), extent);
+
+            if (gradeIfAdded < grade)
+            {
+                grade = gradeIfAdded;
+
+                // Adding this node improves the aspect ratio (nudges it closer to 1, aka makes it more like
+                // a square) so we should append it to the current row

```

Comment:
Does this mean the aspect ratio of the whole tree map also changes at runtime? I'm curious on how it looks while doing lots of allocations.

Reply:
We actually don't do any allocations at runtime for the most part, the exception is the tooltip nodes there (which I will fix thanks for the catch).

Reply:
So to explain this code a bit more, this code runs when either:

1. The nodes in the tree change or are replaced
2. A gui interaction changes a filter or otherwise modifies what nodes will be visible or not
3. The UI is resized or moved

The layout algorithm itself is then performed on-frame, following this paper: https://www.win.tue.nl/~vanwijk/stm.pdf

The specific line you are commenting on is a heuristic evaluated during layout construction, where we do a sort of greedy assignment of nodes to rows/columns provided that the worst-cast node aspect ratio on that row/column improves (becomes more square). If it gets worse, we start a new row and continue with the remaining area in the parent node. Then we keep descending downwards breadth-first.